src/cdk/datepicker/datepicker-input.ts

```diff
@@ -0,0 +1,361 @@
+
+  /** Emits when the value changes (either due to user input or programmatic change). */
+  _valueChange = new EventEmitter<D | null>();
+
+  /** Emits when the disabled state has changed. */
+  _disabledChange = new EventEmitter<boolean>();
+
+  /** Implemented as part of ControlValueAccessor. */
+  _onTouched = () => {};
+
+  /** Implemented as part of ControlValueAccessor. */
+  protected _controlValAccOnChange: (value: any) => void = () => {};
+
+  /** Implemented as part of ControlValueAccessor. */
+  protected _validatorOnChange = () => {};
+
+  /** Implemented for datepicker CDK subscription. */
+  private _cdkDatepickerSubscription = Subscription.EMPTY;
+
+  /** Implemented for datepicker locale subscription. */
+  private _localeSubscription = Subscription.EMPTY;
+
+  /** The native input element to which this directive is attached. */
+  protected _inputElement: HTMLInputElement;
+
+  /** The form control validator for whether the input parses. */
+  private _parseCdkValidator: ValidatorFn = (): ValidationErrors | null => {
+    return this._lastValueValid ?
+        null : {'cdkDatepickerParse': {'text': this._elementRef.nativeElement.value}};
+  }
+
+  /** The form control validator for the min date. */
+  private _minCdkValidator: ValidatorFn = (control: AbstractControl): ValidationErrors | null => {
+    const controlValue = this._getValidDateOrNull(this._dateAdapter.deserialize(control.value));
+    return (!this.min || !controlValue ||
+        this._dateAdapter.compareDate(this.min, controlValue) <= 0) ?
+        null : {'cdkDatepickerMin': {'min': this.min, 'actual': controlValue}};
+  }
+
+  /** The form control validator for the max date. */
+  private _maxCdkValidator: ValidatorFn = (control: AbstractControl): ValidationErrors | null => {
+    const controlValue = this._getValidDateOrNull(this._dateAdapter.deserialize(control.value));
+    return (!this.max || !controlValue ||
+        this._dateAdapter.compareDate(this.max, controlValue) >= 0) ?
+        null : {'cdkDatepickerMax': {'max': this.max, 'actual': controlValue}};
+  }
+
+  /** The form control validator for the date filter. */
+  private _filterCdkValidator: ValidatorFn = (control: AbstractControl):
+      ValidationErrors | null => {
+    const controlValue = this._getValidDateOrNull(this._dateAdapter.deserialize(control.value));
+    return !this._dateFilter || !controlValue || this._dateFilter(controlValue) ?
+        null : {'cdkDatepickerFilter': true};
+  }
+
+  /** The combined form control validator for this input. */
+  private _cdkValidator: ValidatorFn | null =
+      Validators.compose(
+          [this._parseCdkValidator, this._minCdkValidator, this._maxCdkValidator,
+              this._filterCdkValidator]);
+
+  /** Whether the last value set on the input was valid. */
+  protected _lastValueValid = false;
+
+  /** Element reference associated with this input. */
+  protected _elementRef: ElementRef;
```

Comment:
Same comment on protected member naming, do we need `inputElement` and this? they pretty much refer to the same thing

Reply:
I think we need `elementRef` to initialize the `inputElement`. However, I think putting it in the constructor as a private variable is sufficient.

Reply:
We don't need to save both as member variables though